torch/csrc/api/include/torch/nn/pimpl-inl.h

```diff
@@ -45,3 +45,30 @@ struct is_module_holder_of : is_module_holder_of_impl<
                                  is_module_holder<T>::value,
                                  decay_t<T>,
                                  decay_t<C>> {};
+
+// A collection of templates that allow deducing the return type of the
+// `forward()` method, but only if a module actually has a `forward()` method,
+// and otherwise deduces to the type `void`.
+
+template <bool has_forward_value, typename C, typename... Args>
+struct return_type_of_forward_impl;
+
+template <typename C, typename... Args>
+struct return_type_of_forward_impl<true, C, Args...> {
+  using type = decltype(::std::declval<C>().forward(::std::declval<Args>()...));
```

Comment:
not sure if this handles functions with non-const lvalue references correctly since with `declval`, you always get rvalue references (you can convert rvalue references to const lvalue references, but not to non-const ones).

A better approach might be to use `std::result_of`, there I'd have a higher confidence that they implemented it correctly.

Reply:
Also, please add some `static_assert` test cases for this functionality. And make sure you test different kinds of references.

Reply:
`std::result_of` seems not recommended according to https://stackoverflow.com/questions/2689709/difference-between-stdresult-of-and-decltype